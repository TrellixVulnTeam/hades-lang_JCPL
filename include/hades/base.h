#ifndef HADES_BASE_H
#define HADES_BASE_H

#include "hades/base/sys.h"
#include "hades/base/data.h"

namespace hades {

#define HADES_DEFAULT_COPY(T) T(const T&) = default; \
  auto operator=(const T&) -> T& = default;
#define HADES_DEFAULT_MOVE(T) T(T&&) = default; \
  auto operator=(T&&) -> T& = default;

#define HADES_DELETE_COPY(T) T(const T&) = delete; \
  auto operator=(const T&) -> T& = delete;

template <typename T>
constexpr bool is_copyable_v = std::is_copy_constructible_v<T> && std::is_copy_assignable_v<T>;
template <typename T>
constexpr bool is_movable_v = std::is_move_constructible_v<T> && std::is_move_assignable_v<T>;

} // namespace hades


#endif