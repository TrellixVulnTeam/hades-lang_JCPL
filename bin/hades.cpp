#include "llvm/IR/LLVMContext.h"
#include "hades/core/Context.h"

auto main() -> int {
    hades::core::Context ctx;
    ctx.run();
    return 0;
}