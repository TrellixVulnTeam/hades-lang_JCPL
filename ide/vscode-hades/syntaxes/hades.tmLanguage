<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
    <key>$schema</key>
    <string>https://raw.githubusercontent.com/martinring/tmlanguage/master/tmlanguage.json</string>
    <key>name</key>
    <string>Hades</string>
    <key>patterns</key>
    <array>
      <dict>
        <key>include</key>
        <string>#declarations</string>
      </dict>
    </array>
    <key>fileTypes</key>
    <array>
      <string>hds</string>
    </array>
    <key>repository</key>
    <dict>
      <key>declarations</key>
      <dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#extern-function-def</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#function-def</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#import-as</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#struct-def</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#const-def</string>
          </dict>
        </array>
      </dict>
      <key>extern-function-def</key>
      <dict>
        <key>begin</key>
        <string>\b(extern)\s+</string>
        <key>end</key>
        <string>;</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>storage.modifier.hades</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#function-signature</string>
          </dict>
        </array>
      </dict>
      <key>function-signature</key>
      <dict>
        <key>begin</key>
        <string>\b(def)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>[=;}]</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.function.hades</string>
          </dict>
        </dict>
      </dict>
      <key>function-def</key>
      <dict>
        <key>begin</key>
        <string>\b(def)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>(?&lt;=[}])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.function.hades</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#function-def-signature</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#block</string>
          </dict>
        </array>
      </dict>
      <key>function-def-signature</key>
      <dict>
        <key>begin</key>
        <string>(?&lt;=(def)\s+([a-zA=Z_][a-zA-Z0-9_]*))</string>
        <key>end</key>
        <string>(?=[{])</string>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#function-def-signature-param-list</string>
          </dict>
        </array>
      </dict>
      <key>function-def-signature-param-list</key>
      <dict>
        <key>begin</key>
        <string>(\()</string>
        <key>end</key>
        <string>\)</string>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#param</string>
          </dict>
        </array>
      </dict>
      <key>param</key>
      <dict>
        <key>begin</key>
        <string>\b([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>(?=[,)])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>variable.parameter</string>
          </dict>
        </dict>
      </dict>
      <key>import-as</key>
      <dict>
        <key>begin</key>
        <string>\b(import)\s+([a-zA-Z_][a-zA-Z0-9]*)\b(\.[a-zA-Z_][a-zA-Z0-9_]*)*\s+(as)\s+([a-zA-Z_][a-zA-Z0-9_]*)</string>
        <key>end</key>
        <string>[;
]</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.class.hades</string>
          </dict>
          <key>3</key>
          <dict>
            <key>name</key>
            <string>entity.name.class.hades</string>
          </dict>
          <key>4</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>5</key>
          <dict>
            <key>name</key>
            <string>variable.name</string>
          </dict>
        </dict>
      </dict>
      <key>struct-def</key>
      <dict>
        <key>begin</key>
        <string>\b(struct)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>(?&lt;=[}])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.hades</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.class</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#keywords</string>
          </dict>
        </array>
      </dict>
      <key>block</key>
      <dict>
        <key>begin</key>
        <string>([{])</string>
        <key>end</key>
        <string>([}])</string>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#statements</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#if-statement</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#block</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>statements</key>
      <dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#val-statement</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#return-statement</string>
          </dict>
        </array>
      </dict>
      <key>val-statement</key>
      <dict>
        <key>begin</key>
        <string>\b(val)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>([;
])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>variable.name</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>return-statement</key>
      <dict>
        <key>begin</key>
        <string>\b(return)\b</string>
        <key>end</key>
        <string>([;
])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.control</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>if-statement</key>
      <dict>
        <key>begin</key>
        <string>\b(if)\b</string>
        <key>end</key>
        <string>(?&lt;=[}])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword.control.hades</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#block</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>expressions</key>
      <dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#boolean-literals</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#string-literal</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#variable-expressions</string>
          </dict>
        </array>
      </dict>
      <key>boolean-literals</key>
      <dict>
        <key>match</key>
        <string>\b(true|false)\b</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>constant.language</string>
          </dict>
        </dict>
      </dict>
      <key>string-literal</key>
      <dict>
        <key>begin</key>
        <string>((b?)(["]))</string>
        <key>end</key>
        <string>(["])</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>emphasis</string>
          </dict>
          <key>3</key>
          <dict>
            <key>name</key>
            <string>string.quoted.double</string>
          </dict>
        </dict>
        <key>captures</key>
        <dict>
          <key>0</key>
          <dict>
            <key>name</key>
            <string>string.quoted.double.hades</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#inside-string</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#string-escapes</string>
          </dict>
        </array>
      </dict>
      <key>inside-string</key>
      <dict>
        <key>match</key>
        <string>([^"\\]+)</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>string.quoted.double.hades</string>
          </dict>
        </dict>
      </dict>
      <key>string-escapes</key>
      <dict>
        <key>match</key>
        <string>(\\[n\\t"])</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>constant.character.escape</string>
          </dict>
        </dict>
      </dict>
      <key>keywords</key>
      <dict>
        <key>match</key>
        <string>\b(val)\b</string>
        <key>name</key>
        <string>keyword</string>
      </dict>
      <key>const-def</key>
      <dict>
        <key>begin</key>
        <string>\b(const)\s+([a-zA-Z_][a-zA-Z0-9_]*)\b</string>
        <key>end</key>
        <string>[;
]</string>
        <key>beginCaptures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>keyword</string>
          </dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>variable</string>
          </dict>
        </dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#expressions</string>
          </dict>
        </array>
      </dict>
      <key>variable-expressions</key>
      <dict>
        <key>patterns</key>
        <array>
          <dict>
            <key>include</key>
            <string>#variable-method</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#variable-call</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#variable-lowercase</string>
          </dict>
          <dict>
            <key>include</key>
            <string>#variable-uppercase</string>
          </dict>
        </array>
      </dict>
      <key>variable-lowercase</key>
      <dict>
        <key>match</key>
        <string>\b([a-z_][a-zA-Z0-9_]*)\b</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>variable.name</string>
          </dict>
        </dict>
      </dict>
      <key>variable-uppercase</key>
      <dict>
        <key>match</key>
        <string>\b([A-Z_][a-zA-Z0-9_]*)\b</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>entity.name.class</string>
          </dict>
        </dict>
      </dict>
      <key>variable-call</key>
      <dict>
        <key>match</key>
        <string>\b([a-zA-Z_][a-zA-Z0-9_]*)\(\b</string>
        <key>captures</key>
        <dict>
          <key>1</key>
          <dict>
            <key>name</key>
            <string>entity.name.function</string>
          </dict>
        </dict>
      </dict>
      <key>variable-method</key>
      <dict>
        <key>match</key>
        <string>\b(\.)(([a-zA-Z_][a-zA-Z0-9_]*))\(\b</string>
        <key>captures</key>
        <dict>
          <key>2</key>
          <dict>
            <key>name</key>
            <string>entity.name.function</string>
          </dict>
        </dict>
      </dict>
    </dict>
    <key>scopeName</key>
    <string>hades</string>
  </dict>
</plist>