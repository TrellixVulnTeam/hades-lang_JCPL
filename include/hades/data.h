#ifndef HADES_DATA_H
#define HADES_DATA_H

#include "memory"
#include "string"
#include "string_view"
#include "vector"
#include "llvm/ADT/SmallVector.h"

namespace hades {
using String = std::string;
using StringView = std::string_view;
using StringRef = llvm::StringRef;
using Twine = llvm::Twine;

template <typename T, typename Alloc = std::allocator<T>>
using Vec = std::vector<T, Alloc>;

template <typename T, unsigned Size>
using SmallVec = llvm::SmallVector<T, Size>;
} // namespace hades

#endif
