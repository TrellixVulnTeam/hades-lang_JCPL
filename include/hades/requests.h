//
// Created by dhruv on 03/08/20.
//

#ifndef HADES_REQUESTS_GETBUILDFLAGSREQUEST_H
#define HADES_REQUESTS_GETBUILDFLAGSREQUEST_H

#include "hades/ast/SourceFile.h"
#include "hades/core/CommandLineFlags.h"
#include "llvm/IR/LLVMContext.h"

namespace hades::requests {

template <typename Result>
class Request {
};

struct GetBuildFlagsRequest: public Request<Result<CommandLineFlags, FlagParseError>> {
  const Vec<String>* args;
  GetBuildFlagsRequest(const Vec<String>* args): args{args} {}
};

using ExitCode = uint32_t;
struct BuildObjectFileRequest: public Request<ExitCode> {
  const Vec<String>* args;
  BuildObjectFileRequest(const Vec<String>* args): args(args) {}
};

struct GetParsedSourceFileRequest: public Request<SourceFile> {
  const fs::path* path;
};

struct CodegenRequest: public Request<llvm::Module> {};

} // namespace hades::requests

#endif // HADES_REQUESTs_GETBUILDFLAGSREQUEST_H
