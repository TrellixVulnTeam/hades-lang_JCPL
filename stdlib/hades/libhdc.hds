/// Compiler provided runtime library, used to wrap low level
/// functions provided by the platform's libc, or the OS itself.

/// Opaque file handle returned by libc
struct FILE {}

extern def get_stderr(): *mut FILE = _hdc_get_stderr
extern def get_stdout(): *mut FILE = _hdc_get_stdout
extern def get_stdin(): *mut FILE = _hdc_get_stdin

/// A null terminated string, commonly used by libc
/// apis
type CStr = *u8

/// Mutable version of `MutCStr`
type MutCStr = *mut u8

extern def file_put_cstr(*mut FILE, CStr): Void = _hdc_file_put_cstr
extern def file_put_void_ptr(*mut FILE, *Void): Void = _hdc_file_put_void_ptr
extern def file_put_usize(*mut FILE, usize): Void = _hdc_file_put_usize
extern def file_put_u8(*mut FILE, u8): Void = _hdc_file_put_u8
