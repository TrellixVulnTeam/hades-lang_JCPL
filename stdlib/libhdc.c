#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <memory.h>

#define HDC_NULLABLE_PTR(T) T*
#define HDC_NULLABLE_CONST_PTR(T) const T*
#define CStr = const char*

extern FILE* _hdc_get_stderr() {
    return stderr;
}

extern FILE* _hdc_get_stdout() {
    return stdout;
}

extern FILE* _hdc_get_stdin() {
    return stdin;
}

extern void _hdc_panic(const char* message) {
    fprintf(stderr, "PANIC: %s(%d):%s\n", __FILE__, __LINE__, message);
    abort();
}

extern HDC_NULLABLE_PTR(FILE) _hdc_file_open(const char* filename, const char* mode) {
    return fopen(filename, mode);
}

extern void _hdc_file_put_cstr(FILE* stream, const char* cstr) {
    fputs(cstr, stream);
}

#define DEFINE_FILE_PUT(name, ctype, formatstr)\
    extern void _hdc_file_put_##name(FILE* stream, ctype value) {\
        fprintf(stream, formatstr, value);\
    }

DEFINE_FILE_PUT(usize, size_t, "%lu")
DEFINE_FILE_PUT(u8, uint8_t, "%u")
DEFINE_FILE_PUT(i8, int8_t, "%d")
DEFINE_FILE_PUT(u16, uint16_t, "%u")
DEFINE_FILE_PUT(i16, int16_t, "%d")
DEFINE_FILE_PUT(u32, uint32_t, "%u")
DEFINE_FILE_PUT(i32, int32_t, "%d")
DEFINE_FILE_PUT(u64, uint32_t, "%u")
DEFINE_FILE_PUT(i64, int32_t, "%d")
DEFINE_FILE_PUT(f32, float, "%f")
DEFINE_FILE_PUT(f64, double, "%f")
DEFINE_FILE_PUT(void_ptr, void*, "%x")

#undef DEFINE_FILE_PUT

void _hdc_memset(
    void* dest,
    int value,
    size_t size
) {
    memset(dest, value, size);
}

