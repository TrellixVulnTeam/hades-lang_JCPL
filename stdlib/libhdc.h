#ifndef LIB_HDC_H
#define LIB_HDC_H

#endif