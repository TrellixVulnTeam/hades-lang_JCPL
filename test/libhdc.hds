import hades.libhdc.{file_put_cstr, get_stdout}


def main(): Void {
  file_put_cstr(get_stdout(), b"file_put_cstr\n")
}