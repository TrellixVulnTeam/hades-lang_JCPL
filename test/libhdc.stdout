file_put_cstr
10
20